core/opendaq/device/include/opendaq/device_impl.h

```diff
@@ -375,6 +385,44 @@ ErrCode GenericDevice<TInterface, Interfaces...>::unlock(IUser* user)
         }
     }
 
+    if (!this->coreEventMuted && this->coreEvent.assigned())
+        this->triggerCoreEvent(CoreEventArgsDeviceLockStateChanged(false));
+
+    return OPENDAQ_SUCCESS;
+}
+
+template <typename TInterface, typename... Interfaces>
+ErrCode GenericDevice<TInterface, Interfaces...>::isLockedInternal(Bool* locked)
+{
+    OPENDAQ_PARAM_NOT_NULL(locked);
+    *locked = this->userLock.isLocked();
+    return OPENDAQ_SUCCESS;
+}
+
+template <typename TInterface, typename... Interfaces>
+ErrCode GenericDevice<TInterface, Interfaces...>::forceUnlock()
+{
+    std::scoped_lock syncLock(this->sync);
+
+    ErrCode status = forceUnlockInternal();
```

Comment:
Should we return here if device is already unlocked?

Reply:
I would say no. Even if device is unlocked it might still have some locked children which you want to unlock. In this case you can call unlock on the root device and you can be sure that after the call, every device will be unlocked.